### PR TITLE
Store multiple firm IDs

### PR DESCRIPTION
**Pull request by AgustinSilverfin**
The idea is to be able to store multiple firm ids for reconciliations and shared parts. That way we would be able to interact with different environments with the same set of templates (e.g. development, staging, production)

Currently, `config.json` looks like:
```
{
  "id": 123456,
  ...
}
```
From now on, we will store it like this:
```
{ 
  "id": {
    "123": 123456,    // firmId : templateId
  },
 ...
}
``` 
Also, in the `config.json` of the shared part we added two extra parameters (`type` and `handle`) about each template that uses the shared part:
```
{
  ...
  "used_in": [
    { 
      "id":  {
        "123: 123456
      },
      "type": "reconciliation",
      "handle": "reconciliation_handle",
} 
```

It will allow us to search this templates in environments where don't have the corresponding ID yet

This can be considered a breaking change since, when implemented, the existing reference of the id won't be useful anymore. 
Any call to `update-reconciliation` or `update-shared-part` command will fail (showing a message in the terminal indicating the missing ID)

We should run the commands `get-reconciliation-id --all` and `get-shared-part-id --all` in existing repositories to update every template and have the new syntax

One advantage of this new flow is that allow us to replicate all reconciliations, shared parts and relationships between them from one environment into another one. The only missing part now is the command to `create-reconciliation` (which we can already add since the required API endpoint has been implemented)

Edit 02/02/2023: Considering that this will be a breaking change, and some functionality won't be working until you update the ids, I think that we could raise a message when the user is trying to use a command and still have the old syntax, to suggest to update the ids before continuing ? I'll look into this and how can be done

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -174,18 +231,11 @@ async function persistReconciliationText(firmId, handle) {
...
-        ...(await SF.findReconciliationText(firmId, handle)),
-      }.id;
-    }
-    if (!reconciliationTextId) {
-      throw "Reconciliation not found";
+    if (!config || !config.id[firmId]) {
+      console.log(`Reconciliation ${handle}: ID is missing. Aborted`);
+      process.exit(1);
     }
+    let reconciliationTextId = (reconciliationTextId = config.id[firmId]);
```
[the remark this answers is not in the document]
Comment (Woetfin) [reply index.js:238]: Should be fixed

File: index.js [hunk 2]
```diff
@@ -213,11 +263,37 @@ async function importExistingSharedPartById(firmId, id) {
     sharedPart.data.text
   );
 
-  let config = {
-    id: sharedPart.data.id,
+  // Adjust ID and find reconciliation handle
+  let used_in = [];
+  for (reconciliation of sharedPart.data.used_in) {
+    // Search in repository
+    let reconHandle = fsUtils.findHandleByID(
```
[the remark this answers is not in the document]
Comment (Woetfin) [reply index.js:270]: Should be fixed

File: fs_utils.js [hunk 3]
```diff
@@ -136,6 +144,20 @@ function getTemplatePaths(relativePath) {
...
 
+// Get the handle/name of a reconciliation/shared part by it's ID
+function findHandleByID(type, id) {
+  if (type !== "shared_parts" && type !== "reconciliation_texts") {
+    throw "type should be shared_parts or reconciliation_texts";
+  }
+  let templatesArray = getTemplatePaths(type);
+  for (let templatePath of templatesArray) {
+    let config = readConfig(templatePath);
+    if (config.id[firmId] === id) {
```
[the remark this answers is not in the document]
Comment (Woetfin) [reply fs_utils.js:155]: Should be fixed

File: fs_utils.js [hunk 4]
```diff
@@ -136,6 +144,20 @@ function getTemplatePaths(relativePath) {
   return templatesArray;
 }
 
+// Get the handle/name of a reconciliation/shared part by it's ID
+function findHandleByID(type, id) {
```
[the remark this answers is not in the document]
Comment (Woetfin) [reply fs_utils.js:148]: Added it in the latest commit

**Review by Woetfin — CHANGES_REQUESTED** [review]
I tested each command and added some fixes for the things that weren't working in the cases I tested, can you review if they're ok for you? 
1. `import-reconciliation`:
	1. Existing firmIDs of other firms in the config were overwritten when importing a reconciliation
2. `update-reconciliation`
	1. `let reconciliationTextId = (reconciliationTextId = config.id[firmId])` was being assigned to itself before it was initialized
3. `import-shared-part`
	1. `findHandleByID`: firmId was undefined
	2. Existing ids from other firms were being overwritten in the id property of the shared part config + in the id property of the reconciliation in the `used_in` array
	3. Cover case where config file or `used_in` array did not exist yet
4. `create-shared-part`
	1. Also ensure `shared_parts` folder is created
	2. Create liquid file with minimal content otherwise the shared part is not being created by the API
	3. Removed the 'yes' option since the function already has hardcoded logic to avoid overwriting existing shared parts
5. `add-shared-part`
	1. Cover case where used_in array is empty
6. `remove-shared-part`
	1. Only remove the link to the reconciliation for the firm the command (before the complete reconciliation object was taken out of the array, even if it was still linked to other firms)

File: index.js [hunk 5]
```diff
@@ -213,11 +263,37 @@ async function importExistingSharedPartById(firmId, id) {
...
+      let reconciliationText = await SF.findReconciliationTextById(
+        firmId,
+        reconciliation.id
+      );
+      if (reconciliationText) {
+        reconciliation.handle = reconciliationText.handle;
+      }
+    }
+    reconId = reconciliation.id;
+    reconciliation.id = {};
```
[the remark this answers is not in the document]
Comment (AgustinSilverfin) [reply index.js]: @Woetfin the new logic implemented won't remove the ids with the old format. Should we address this? 

We could get something like this:
```
"used_in": [
    {
      "id": 6379437,
      "type": "reconciliation"
    },
    {
      "id": 6379269,
      "type": "reconciliation"
    },
    {
      "id": 57823969,
      "type": "reconciliation"
    },
    {
      "id": {
        "998": 6379437,
        "11710": 58053574
      },
      "type": "reconciliation",
      "handle": "cedric_handle"
    },
    {
      "id": {
        "998": 6379269
      },
      "type": "reconciliation",
      "handle": "dev_recon_div"
    },
    {
      "id": {
        "998": 57823969,
        "11710": 58029422
      },
      "type": "reconciliation",
      "handle": "repo_test"
    }
  ]
```

[the remark this answers is not in the document]
Comment (AgustinSilverfin) [reply index.js]: Added snippet to remove old formatted ids

**Review by Woetfin — APPROVED** [review]
Tested all commands and works as expected for me

**Pull request merged**